Processing/yOCTFindTissueSurface.m

```diff
@@ -27,9 +28,91 @@
+
+    for x = 1:x_size % Iterate over each X position within the slice
+        found = false;  % Initialize flag to track if surface is found
+        % Option to decrease confirmations needed if surface is not found
+        
+        for confirmations_required = base_confirmations_required:-1:11 % Adjustable decrease
+            if found
+                break;  % Exit early if surface is found
+            end
+
+            for z = start_depth:z_size
+                if current_slice(z, x) >= intensity_threshold % Check if pixel intensity is above threshold
+
+                    % Confirm surface if a sufficient number of consecutive pixels meet the criteria
+                    if (z + confirmations_required <= z_size) && all(current_slice(z+1:z+confirmations_required, x) > 0)
+                        surface_column(x) = z;  % Record the surface depth
+                        found = true;  % Update flag
+                        break; % Stop scanning as surface is confirmed
+                    end
+                end
+            end
+        end
+    end
+
+    surface_depth(:, y) = surface_column;  % Store the detected surface depths
+end
+
+% Configure the Gaussian Kernel Smoothing
+sigma = 1.5;  % Deviation for the Gaussian kernel
+kernel_size = ceil(sigma * 3) * 2 + 1;  % Determine kernel size based on sigma
+gaussian_kernel = fspecial('gaussian', [kernel_size, kernel_size], sigma);  % Create Gaussian kernel
+
+% Preprocessing Before Smoothing
+surface_depth(isnan(surface_depth)) = 0;  % Replace NaN values with zeros to prepare for smoothing
+smoothed_surface_depth = conv2(surface_depth, gaussian_kernel, 'same');  % Apply Gaussian filter to smooth the surface depth map
+
+% Normalize the Smoothed Data
+normalizing_kernel = conv2(surface_depth ~= 0, gaussian_kernel, 'same');  % Create a normalization matrix from non-zero entries
+smoothed_surface_depth = smoothed_surface_depth ./ normalizing_kernel;  % Normalize to account for initial zeros used for NaNs
+smoothed_surface_depth(normalizing_kernel == 0) = NaN;  % Restore NaN values where no original data existed
+smoothed_surface_depth = round(smoothed_surface_depth);  % Round smoothed values to the nearest integer for uniform depth representation
+
+% Assign z, x, y, coordinates
+z = dim.z.values(:); x = dim.x.values(:); y = dim.y.values(:);
+
+% Convert Depth found from Pixels to Units in dim.x.units like microns or mm
+surface_depth_pixels = smoothed_surface_depth.'; % Transpose for consistent global orientation system
+unit_per_pixel = abs(z(2) - z(1)); % Calculate Z unit per pixel ratio
+surfacePosition = surface_depth_pixels * unit_per_pixel; % Get Surface Depth in Units
+
+
+%% Generate Heatmap of Surface Height for user visualization
+if isVisualize
+    figure;
+    imagesc(x, y, surfacePosition);
+    colormap(flipud(jet));
+    colorbar;
+    set(gca, 'YDir', 'normal');
+    title(['Surface position in ', dim.x.units, ' (View from the top)']);
```

Comment:
Surface position should be in dim.z.units (which may or may not be the same as dim.x.units)